Argument buffers and Metal bump mapping support

colincornaby opened the pull request:
This PR implements argument buffers in the Metal renderer - and implements them first through bump mapping. Eventually I would like argument buffers to replace the entire layer system in the Metal shaders. I'm actively working on that - will be a future PR. If anyone is interested in rendering - please take your time on this one. Any improvements/feedback here will be very helpful for future work.

This came about because @dpogue asked about multiple bump maps in the original PR. There aren't an unlimited number of blind points in the shaders and the bind points have to be predefined - so supporting multiple bump maps would have been complicated.

Argument buffers replace bind points in shaders. I bind resources to the argument buffers instead, and then pass the buffers to the shader. The advantage is that now the bind points for textures, samplers, and other resources don't need to be predefined in the shader. And because they don't need to be predefined I can pass an array of argument buffers of any length and the shader can iterate over them. That means I can effectively bind an unlimited number of resources within shaders. This solves the problem of how we deliver an indeterminate amount of bump maps to the shader.

There is also a performance pickup. I can leave the argument buffers on the GPU, and instead of having to bind textures and samplers individually, I can just rebind the buffer I have cached. I have to be careful doing this around features like push over layers and piggyback layers. I'm doing some state tracking and comparisons right now to guard against changes and update the argument buffers when necessary. _This will get messier when layers themselves move over._

While this change lets us effectively support an unlimited number of bump maps - the structure passed between the vertex and fragment stages is still compile time defined. This has constrained the number of bump maps that can be supported by the shader to 8. (Which seems totally reasonable to me.)

Argument buffers work differently in Metal 3 and Metal 2. This code supports both approaches.

- Tier 2 argument buffers (Metal 3) give direct access to the underlying buffer. I can directly store texture/sampler/etc addresses into those buffers.
- Tier 1 argument buffers (Metal 2) have to support hardware with an inconsistent memory layout. An encoder object must be created that manages those differences - which adds some overhead.

Not all contributor Macs will support Metal 3, Apple lists the Metal 3 requirements here:
https://support.apple.com/en-us/102894

Tier 1 and Tier 2 buffers are also a little different in the shader. Tier 1 buffers still need ids for each binding so the encoder knows what is being bound where. Tier 2 argument buffers give me direct access - so there is no need to binding ids.

All previous conversations about the quality of the bump map still apply here. The code is present for full quality bump maps. But we're emulating the D3D Plasma behavior of ignoring the "height" component of the normal map.

colincornaby committed:
Metal bump mapping with arguement buffers

colincornaby committed:
Setting sampler for bump map

colincornaby committed:
Cleaning up plMetalArgumentBuffer member names

Hoikas commented on Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalArgumentBuffer.cpp:
```diff
@@ -0,0 +1,140 @@
+        return;
+    }
+    ConfigureBuffer();
+    int i = 0;
```
If you need an index, we should probably not be using a range based for. Instead, do an old fashioned `for (int = 0; i < bumps.size(); ++i)` and get the vector element you need.

Hoikas commented on Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalPipelineState.h:
```diff
@@ -148,12 +148,18 @@ struct plMetalFragmentShaderDescription
+                                    memcmp(fBlendModes, p.fBlendModes, sizeof(fBlendModes)) == 0 &&
+                                    memcmp(fMiscFlags, p.fMiscFlags, sizeof(fMiscFlags)) == 0 &&
+                                    fUsePerPixelLighting == p.fUsePerPixelLighting &&
+                                    fNumBumpMaps == p.fNumBumpMaps;
```
This is so long that it wraps in split diff mode.

Hoikas commented on Sources/Plasma/FeatureLib/pfMetalPipeline/ShaderSrc/FixedPipelineShaders.metal:
```diff
@@ -449,9 +499,48 @@ half4 FragmentShaderArguments::sampleLayer(const size_t index, const half4 verte
+    constexpr bool performBaseLighting = bumpMapIsAdditive || !bumpMap;
+    
+    if (performBaseLighting) {
+        lightingContributionColor = perPixelLighting ? calcLitMaterialColor(lighting, in.vtxColor, materialLighting, in.worldPos, in.normal) : in.vtxColor;
```
This line is so long it wraps in split diff mode.

Hoikas commented on Sources/Plasma/FeatureLib/pfMetalPipeline/ShaderSrc/FixedPipelineShaders.metal:
```diff
@@ -280,10 +324,16 @@ vertex ColorInOut pipelineVertexShader(Vertex in [[stage_in]],
+    
+    for (int bumpMap=0; bumpMap<numBumpMaps; bumpMap++) {
+        (&out.T)[bumpMap] = normalize(uniforms.localToWorldMatrix * float4((&in.texCoord1)[bumpInfo[bumpMap].dTangentIndex[0]], 0.f)). xyz;
+        (&out.B)[bumpMap] = normalize(uniforms.localToWorldMatrix * float4((&in.texCoord1)[bumpInfo[bumpMap].dTangentIndex[1]], 0.f)). xyz;
```
Wraps in split diff

colincornaby committed:
Incorperating feedback from code review

colincornaby committed:
Fixing one more loop from code review

Hoikas commented on Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalMaterialShaderRef.cpp:
```diff
@@ -301,6 +322,63 @@ void plMetalMaterialShaderRef::ILoopOverLayers()
+    bumpMapping.texture = static_cast<plMetalTextureRef*>(bumpTextureLayer->GetTexture()->GetDeviceRef())->fTexture;
+    MTL::SamplerState* samplerState = fPipeline->fDevice.SampleStateForClampFlags(hsGMatState::hsGMatClampFlags(bumpTextureLayer->GetClampFlags()));
+    bumpMapping.sampler = samplerState;
+    bumpsOut.emplace_back(bumpMapping);
```
It would be nice if we were doing `plMetalBumpMapping& bumpMapping = bumpsOut.emplace_back()` so that the bump mapping object is only constructed once.

colincornaby committed:
Adding sampler ID to bump mapping argument descriptor

This shouldn’t have worked before on Metal 2 class hardware. Need to do a retest on older hardware.

colincornaby committed:
Fixing Tier 1 bumpmaps

Now tested

Hoikas commented on Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalArgumentBuffer.cpp:
```diff
@@ -0,0 +1,142 @@
+    return array;
+}
+
+plMetalBumpArgumentBuffer::plMetalBumpArgumentBuffer(MTL::Device* device, size_t numElements) : plMetalArgumentBuffer<plMetalBumpmap>(device, numElements)
```
```suggestion
plMetalBumpArgumentBuffer::plMetalBumpArgumentBuffer(MTL::Device* device, size_t numElements)
    : plMetalArgumentBuffer<plMetalBumpmap>(device, numElements)
```

I'm not sure how I missed this before

colincornaby committed:
Cutting over to new argument buffer tier query

Adds support for running Metal 2 argument buffers on Metal 3 Mac

colincornaby committed:
Address feedback from code review

Hoikas commented on Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalArgumentBuffer.h:
```diff
@@ -0,0 +1,185 @@
+    {
+        fBuffer[0]->release();
+        fBuffer[1]->release();
+        fBuffer[2]->release();
```
```suggestion
        for (auto& buffer : fBuffer) {
            if (buffer)
                buffer->release();
        }
```

C style arrays can use range based for loops. Further, each element of `fBuffer` is initialized to `nullptr`, so we need to be cautious about dereferencing them.

colincornaby replied:
Same feedback here. I can implement the loop, but in since this is Obj-C, it's ok to send messages/call functions on nullptr.

Hoikas replied:
My purpose for suggesting the loop is to avoid hardcoding the existence of specifically three buffers.

Hoikas commented on Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalArgumentBuffer.h:
```diff
@@ -0,0 +1,185 @@
+        fBuffer[0]->release();
+        fBuffer[1]->release();
+        fBuffer[2]->release();
+        fEncoder->release();
```
```suggestion
        if (fEncoder)
            fEncoder->release();
```

This is initialized to `nullptr`.

Hoikas replied:
That sounds like we'd be depending on an implementation detail, and I can't say I'm a fan of that.

colincornaby replied:
Hey @Hoikas - I've been thinking about this conversation and I had a realization. Please check out the last commit. I think it should resolve the issues you had here.

We're sort of deadlocked on the retain/release thing. But I remembered that Metal-cpp has a NS::SharedPtr type. It adds C++ SharedPtr semantics to Obj-C types with no extra overhead. That _should_ resolve the issue. I don't have to manually release the objects any more - so there is not more discussion on potentially sending a release to a nullptr. It also simplifies the code.

NS::SharedPtr wasn't originally part of Metal-cpp, so it wasn't originally adopted by this renderer. I think we should adopt it now.

There is no NS::UniquePtr - so I'll have to ask for a bit of understanding there. But all Obj-C pointer types already have SharedPtr mechanics built in - so there wouldn't be a point in NS::UniquePtr.

Hoikas replied:
Yes, this looks a lot better. I see a few very, very, very minor things left, otherwise this will be a 👍 from me. I'd like for @dpogue to look behind me, though, if possible. My mac and Metal knowledge is still quite poor.

Hoikas commented on Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalArgumentBuffer.h:
```diff
@@ -0,0 +1,185 @@
+    // Subclasses must override this to create an encoder for
+    // tier 1 buffers. Encoder creation needs to know the
+    // buffer layout.
+    virtual NS::Array*        GetArgumentDescriptors() = 0;
```
This looks like it should probably be a `const` method.

Hoikas commented on Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalArgumentBuffer.h:
```diff
@@ -0,0 +1,185 @@
+
+    void ConfigureBuffer()
+    {
+        fCurrentBufferIndex = (fCurrentBufferIndex + 1) % 3;
```
Instead of hardcoding the 3, should we do something like


```suggestion
        fCurrentBufferIndex = (fCurrentBufferIndex + 1) % std::size(fBuffer);
```

Hoikas commented on Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalArgumentBuffer.h:
```diff
@@ -0,0 +1,185 @@
+    uint8_t             dTangentVIndex;
+    float               scale;
+    MTL::Texture*       texture;
+    MTL::SamplerState*  sampler;
```
Fields should begin with `f`

colincornaby commented on Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalArgumentBuffer.h:
```diff
@@ -0,0 +1,185 @@
+            case plMetalArgumentBufferTier::Tier2:
+                ConfigureTier2();
+                break;
+#endif
```
I'm being careful here because if the Metal 3 SDK is not available these symbols _may_ not be available to the compiler. Technically the Metal header will dynamically look up symbols blah blah blah... but Tier 2 symbols won't technically be available to the compiler.

I could require everyone build against the Metal 3 SDK but that will move the minimum required build toolchain to Xcode 14.

Hoikas replied:
If you declare dummy values, what the SDK provides is irrelevant.

colincornaby committed:
Feedback from code review on argument buffers

colincornaby commented:
I'm working through clearing my pull requests. This one is important because it finalizes bump mapping in the Metal renderer. It's also holding up another PR with performance improvements - getting concerned I'll lose my mental context on the changes I made in that PR. Hoping we can get this one through soon.

colincornaby committed:
Migrating owned pointes in Metal argument buffers to NS::SharedPtr

Hoikas commented on Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalArgumentBuffer.cpp:
```diff
@@ -0,0 +1,143 @@
+    ConfigureBuffer();
+    if (fTier == plMetalArgumentBufferTier::Tier1) {
+        // Tier 1, because Tier 1 doesn't guarantee memory layout, go through the encoder
+        for (int i = 0; i < bumps.size(); ++i) {
```
These two for loops - should the `i`ndex be typed `size_t` instead of `int`?

Hoikas commented on Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalArgumentBuffer.h:
```diff
@@ -0,0 +1,174 @@
+    // Tier of argument buffers targeted by this instance
+    plMetalArgumentBufferTier fTier;
+    // Current buffer in the triple buffer rotation
+    size_t                      fCurrentBufferIndex;
```
It would be nice to move these members up to the top.

colincornaby committed:
Applying feedback from code review
